lib/tbot/ssh_proxy.go

```diff
@@ -0,0 +1,262 @@
+		conn, err = resumption.WrapSSHClientConn(ctx, conn, func(ctx context.Context, hostID string) (net.Conn, error) {
+			// if the connection is being resumed, it means that
+			// we didn't need the agent in the first place
+			var noAgent agent.ExtendedAgent
+			conn, _, err := pclt.DialHost(ctx, net.JoinHostPort(hostID, "0"), cluster, noAgent)
+			return conn, err
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	defer conn.Close()
+
+	stdio := utils.CombineReadWriteCloser(io.NopCloser(os.Stdin), utils.NopWriteCloser(os.Stdout))
+	err = trace.Wrap(utils.ProxyConn(ctx, stdio, conn))
+	if errors.Is(err, context.Canceled) {
+		err = nil
+	}
+
+	return trace.Wrap(err)
+}
+
+// resolveTargetHost determines the ssh instance to be connected to based on either
+// the provided search or query. The auth client is intentionally single use and
+// closed to reduce resource consumption after the host has been resolved. Any future
+// changes that require an additional request to auth should reuse the client instead
+// of creating one per request.
+func resolveTargetHost(ctx context.Context, cfg client.Config, search, query string) (types.Server, error) {
+	apiClient, err := client.New(ctx, cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer apiClient.Close()
+
+	nodes, err := client.GetAllResources[types.Server](ctx, apiClient, &proto.ListResourcesRequest{
+		ResourceType:        types.KindNode,
+		SearchKeywords:      libclient.ParseSearchKeywords(search, ','),
+		PredicateExpression: query,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(nodes) == 0 {
+		return nil, trace.NotFound("no matching SSH hosts found for search terms or query expression")
+	}
+
+	if len(nodes) > 1 {
+		return nil, trace.BadParameter("found multiple matching SSH hosts %v", nodes[:2])
+	}
+
+	return nodes[0], nil
+}
+
+func parseIdentity(botConfig *config.BotConfig, proxy, cluster string) (*identity.Facade, agent.ExtendedAgent, error) {
+	destination, err := tshwrap.GetDestinationDirectory(botConfig)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	identityPath := filepath.Join(destination.Path, config.IdentityFilePath)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	key, err := identityfile.KeyFromIdentityFile(identityPath, proxy, cluster)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	i, err := identity.ReadIdentityFromStore(&identity.LoadIdentityParams{
+		PrivateKeyBytes: key.PrivateKeyPEM(),
+		PublicKeyBytes:  key.MarshalSSHPublicKey(),
+	}, &proto.Certs{
+		SSH:        key.Cert,
+		TLS:        key.TLSCert,
+		TLSCACerts: key.TLSCAs(),
+	})
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	agentKey, err := key.AsAgentKey()
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	keyring, ok := agent.NewKeyring().(agent.ExtendedAgent)
+	if !ok {
+		return nil, nil, trace.BadParameter("unexpected keyring type %T, expected agent.ExtendedKeyring", keyring)
```

Comment:
```suggestion
		return nil, nil, trace.BadParameter("unexpected keyring type %T, expected agent.ExtendedAgent", keyring)
```

```suggestion
		return nil, nil, trace.BadParameter("unexpected keyring type %T, expected agent.ExtendedAgent (this is a bug)", keyring)
```

also note for future improvements: we could probably just pass and receive `agent.Agent`s around, I don't think we ever actively make use of the agent extensions